### PR TITLE
Fixes for julia 1.0

### PR DESCRIPTION
**Pull request by IssamT**
This is a continuation on the nice work started by @matbesancon in #179. Tests are now passing for 0.6 and 1.0. Still some work to do:

- [x] pass conic tests. I temporarily disabled them since MPB still uses the deprecated `speye`.
- [ ] fix `cpx_ccall_intercept`. I temporarily replaced it with `cpx_ccall`.
- [x] fix finalizers.

### CODE REVIEW

**Comment by joaquimg** [comment]
> pass conic tests. I temporarily disabled them since MPB still uses the deprecated speye.
 
Now works, tested in Xpress and Gurobi

File: src/CPLEX.jl [hunk 1]
```diff
@@ -84,6 +85,11 @@ module CPLEX
     #        cplex_model
 
     using Compat
+    
+    if VERSION >= v"0.7.0-DEV.3382"
+        using SparseArrays
+        using LinearAlgebra
+    end
```
Comment (joaquimg) [inline src/CPLEX.jl]: We could also use `Compat.XXX`

**Comment by giadasp** [comment]
Sorry guys, where I can find the updated files that work with Julia 1.0?
I can find only the files with all the changes...
Thank you!

**Comment by giadasp** [comment]
Thank you guys for updating this package, I've still problems with Julia 1.0, in particular julia doesn't find the function `contains` and even if I substitute it with `occursin` it's not able to find the path of CPLEX in my pc.
I also tried by setting ENV["CPLEX_STUDIO_BINARIES"] to the right path but still when I try `using CPLEX` it says:
` CPLEX not properly installed. Please run Pkg.build("CPLEX")
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] top-level scope at C:\Users\User\.julia\packages\CPLEX\FUtHu\src\CPLEX.jl:17
 [3] include at .\boot.jl:317 [inlined]
 [4] include_relative(::Module, ::String) at .\loading.jl:1041
 [5] include(::Module, ::String) at .\sysimg.jl:29
 [6] top-level scope at none:2
 [7] eval at .\boot.jl:319 [inlined]
 [8] eval(::Expr) at .\client.jl:389
 [9] top-level scope at .\none:3
in expression starting at C:\Users\User\.julia\packages\CPLEX\FUtHu\src\CPLEX.jl:14`
What's the problem in your opinion?

File: src/CPLEX.jl (hunk 1 above)
Comment (matbesancon) [reply src/CPLEX.jl]: given the guideline update on [discourse](https://discourse.julialang.org/t/jump-and-julia-0-7-1-0/12964), I think we can remove this all-together and use 1.0-compatible features only
So remove the version check and no Compat

**Comment by giadasp** [comment]
Update:
I restarted Julia, made a lot of changes in JuMP files (since it was not working with Julia 1.0) and now I can start to optimize a model but still there are problems with CPLEX:

- The manifest still not add the required packages "LinQuadOptInterface","SparseArrays" and "LinearAlgebra" --> add them manually (I don't know how to let Julia add them automatically)
- changed `cfunction` to `@cfunction` in CPLEX wrapper files
-`MethodError: no method matching cbgetstate(::CPLEX.CplexLazyCallbackData)` , I triEd to add it in the exported functions list in CplexSolverInterface.jl but it doesn't work, 
what I can do?

**Comment by matbesancon** [comment]
@giadasp I think this is unrelated to this PR, maybe consider getting help on discourse or slack for this
@IssamT I think we need rebasing here

File: src/CPLEX.jl [hunk 2]
```diff
@@ -117,10 +123,12 @@ module CPLEX
     
     # These are undocumented JuMP extensions for CPLEX which
     # will need to be hosted in a separate package for Julia 0.6 and later.
-    if isdir(Pkg.dir("JuMP")) && VERSION < v"0.6-"
-        try
-            eval(current_module(), Expr(:import,:JuMP))
-            include("JuMPfunctions.jl")
-        end
-    end
+    # if isdir(Pkg.dir("JuMP")) && VERSION < v"0.6-"
```
Comment (mlubin) [inline src/CPLEX.jl]: Feel free to delete this chunk of code and remove `JuMPfunctions.jl`.

File: src/CplexSolverInterface.jl [hunk 3]
```diff
@@ -341,18 +342,18 @@ export cbaddboundbranchup!,
...
-setheuristiccallback!(m::CplexMathProgModel,f) = (m.heuristiccb = f)
+MathProgBase.setlazycallback!(m::CplexMathProgModel,f) = (m.lazycb = f)
+MathProgBase.setcutcallback!(m::CplexMathProgModel,f) = (m.cutcb = f)
+MathProgBase.setheuristiccallback!(m::CplexMathProgModel,f) = (m.heuristiccb = f)
 setbranchcallback!(m::CplexMathProgModel,f) = (m.branchcb = f)
 setincumbentcallback!(m::CplexMathProgModel,f) = (m.incumbentcb = f)
-setinfocallback!(m::CplexMathProgModel,f) = (m.infocb = f)
+MathProgBase.setinfocallback!(m::CplexMathProgModel,f) = (m.infocb = f)
 
 function cbgetmipsolution(d::CplexCallbackData)
```
Comment (mlubin) [inline src/CplexSolverInterface.jl]: The the `cb` functions also need a `MathProgBase.` prefix if they're implementing the MPB API.

File: src/CplexSolverInterface.jl [hunk 4]
```diff
@@ -615,10 +616,10 @@ function setmathproglazycallback!(model::CplexMathProgModel)
     set_param!(model.inner.env, "CPX_PARAM_MIPCBREDLP", 0)
     set_param!(model.inner.env, "CPX_PARAM_PRELINEAR", 0)
     set_param!(model.inner.env, "CPX_PARAM_REDUCE", CPX_PREREDUCE_PRIMALONLY)
-    cpxcallback = cfunction(mastercallback, Cint, (Ptr{Void}, Ptr{Void}, Cint, Ptr{Void}, Ptr{Cint}))
+    cpxcallback = cfunction(mastercallback, Cint, (Ptr{Nothing}, Ptr{Nothing}, Cint, Ptr{Nothing}, Ptr{Cint}))
```
Comment (mlubin) [inline src/CplexSolverInterface.jl]: @cfunction

File: src/CplexSolverInterface.jl [hunk 5]
```diff
@@ -362,7 +363,7 @@ end
 
 function cbgetmipsolution(d::CplexCallbackData, sol::Vector{Cdouble})
     @assert d.state == :MIPSol
-    stat = @cpx_ccall(getcallbacknodex, Cint, (Ptr{Void},Ptr{Void},Cint,Ptr{Cdouble},Cint,Cint),
+    stat = @cpx_ccall(getcallbacknodex, Cint, (Ptr{Nothing},Ptr{Nothing},Cint,Ptr{Cdouble},Cint,Cint),
```
Comment (mlubin) [inline src/CplexSolverInterface.jl]: Use `Cvoid` instead of `Nothing` when referring to the C type.

File: src/CplexSolverInterface.jl (hunk 4 above)
Comment (giadasp) [reply src/CplexSolverInterface.jl]: It's still `cfunction` instead of `@cfunction`...

Comment (mlubin) [reply src/CplexSolverInterface.jl]: `cfunction` does not exist in Julia 1.0

Comment (giadasp) [reply src/CplexSolverInterface.jl]: That's why you should change it ...
The code still contains `cfunction`

Comment (mlubin) [reply src/CplexSolverInterface.jl]: That was exactly the point of my comment :)

Comment (giadasp) [reply src/CplexSolverInterface.jl]: Btw I still cannot use `@lazyconstraint`, it gives always the same error:
`MethodError: no method matching cbgetstate(::CPLEX.CplexLazyCallbackData)`

Comment (giadasp) [reply src/CplexSolverInterface.jl]: Sorry Miles, I thought you were making the changes, I didn't understand was only a comment. -.-"
It's sunday, I'm working, please apologize.

File: src/CPLEX.jl (hunk 1 above)
Comment (IssamT) [reply src/CPLEX.jl]: @matbesancon, but those are needed for 0.7 and above.

File: src/cpx_common.jl [hunk 6]
```diff
@@ -22,40 +24,42 @@ end
...
+        # ret
     end
 end
 
 const GChars = Union{Cchar, Char}
 const IVec = Vector{Cint}
 const FVec = Vector{Cdouble}
 const CVec = Vector{Cchar}
-const CoeffMat = Union{Matrix{Cdouble}, SparseMatrixCSC{Cdouble}}
+const CoeffMat = Union{AbstractArray{Cdouble, 2}, SparseMatrixCSC{Cdouble}}
```
Comment (matbesancon) [inline src/cpx_common.jl]: `AbstractMatrix` would be more concise?

**Comment by IssamT** [comment]
@mlubin, did you mean the file `test/callback.jl` in JuMP? I fixed some callback errors but when running the file with julia 1.0 I get some errors in that are due to jump I think.

**Comment by IssamT** [comment]
These are the errors I have now for the test/callback.jl https://justpaste.it/4x67j

**Comment by mlubin** [comment]
@IssamT Could you try https://github.com/JuliaOpt/MathProgBase.jl/pull/227?

**Comment by IssamT** [comment]
Apparently this doesn't work. from JuMP's callback.jl  :

````
julia
@test macroexpand(:(@lazyconstraint(cb, sum(x) <= 0, badkwarg=true))).head == :error
`````

**Comment by mlubin** [comment]
`CPXMIP_ABORT_FEAS` should map to `UserLimit` under MPB.

**Comment by IssamT** [comment]
Thanks. Fixed. So the only issue remaining is the test with `macroexpand` that doesn't run.

**Pull request merged**